docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc

```diff
@@ -386,6 +386,186 @@ It is possible to use multiple expressions in the role definition.
+
+CAUTION: If you plan to use the `@PermissionsAllowed` on the IO thread, review the information in xref:security-proactive-authentication-concept.adoc[Proactive Authentication].
+NOTE: The `@PermissionsAllowed` is not repeatable on class-level due to limitations of Quarkus interceptors.
+Please find well-argued explanation in the xref:cdi-reference.adoc#repeatable-interceptor-bindings[Repeatable interceptor bindings] section of the CDI reference.
+
+On CDI beans, it is also possible to take custom `java.security.Permission` much further.
+Let's image that in your application, you have following service:
+
+[source,java]
+----
+import io.quarkus.security.PermissionsAllowed;
+import io.smallrye.mutiny.Uni;
+import jakarta.enterprise.context.ApplicationScoped;
+
+import java.security.Permission;
+import java.util.Objects;
+
+@ApplicationScoped
+public class GreetingsService {
+
+    @PermissionsAllowed(value = "ignored", permission = HelloPermission.class) <1>
+    public Uni<String> greetings(int index, Object obj, String hello) {
+        return Uni.createFrom().item("Welcome!");
+    }
+
+    @PermissionsAllowed(value = "ignored", permission = HelloPermission.class, params = "greetings") <2>
+    public Uni<String> greetings(String farewell, String greetings) {
+        return Uni.createFrom().item("Welcome!");
+    }
+
+    public static class HelloPermission extends Permission {
+        private final boolean pass;
+
+        public HelloPermission(String name, String[] actions, String greetings) { <3>
+            super(name);
+            this.pass = "hello".equals(greetings);
+        }
+
+        @Override
+        public boolean implies(Permission permission) {
+            return pass;
+        }
+
+        ...
+    }
+
+}
+----
+<1> Formal parameter `hello` is identified as the first `String` parameter and passed to `HelloPermission`.
+However this option comes with a price, as the `HelloPermission` must be instantiated every single time `greetings` method is invoked.
+<2> Here, the first `String` parameter is `farewell`, therefore we marked `greetings` parameter explicitly via `PermissionsAllowed#params`.
+Please note that both constructor and annotated method must have parameter `greetings`, otherwise validation will fail.
+<3> There must be exactly one constructor of a custom `Permission` class, also first parameter is always considered a permission name (must be `String`).
+Optionally, Quarkus may pass Permission actions to the constructor. Just declare the second parameter as `String[]`.
+
+CAUTION: Passing method parameters to a custom `Permission` constructor is not supported on RESTEasy Reactive endpoints,
+because there, security checks are done before serialization.
+
+Currently, there is only one way to add permissions, and that is xref:security-customization.adoc#security-identity-customization[Security Identity Customization].
```

Comment:
Hey @michalvavrik This is a great PR and indeed as you explained earlier more iterations may follow, but it does feel to me that there should be something ready out of the box, for this feature be useful without having to create custom security augmentors.
Can we think of some obvious candidates ? Example, start from extending HTTP security policy configuration to support mapping between roles and list of permissions ? (ex, admin role can write, read while user role can read only, etc) - that would work for all the security extensions.

In the next steps we can handle OIDC scopes etc

Reply:
Thanks @sberyozkin . Well, I absolutely agree, but it will add complexity to this PR. I'd suggest that week after this PR is merged (whenever it will be in the future), I'll have prepared PR for https://github.com/quarkusio/quarkus/issues/12219.

I'm happy to start extending HTTP security policy configuration in this PR, but please confirm you don't mind there will be more changes :-) I'd prefer to do it afterwards.

Reply:
@michalvavrik You are right, makes sense to do it in the next PR and let the current PR code settle for a bit in meantime